### ISSUE TITLE
New pyro distributions

### ISSUE THREAD

**Issue by VargEM** [pyro-ppl/pyro#2836]
Can Johnson SU and Johnson SB distributions be implemented for pyro??

**Comment by fritzo** [comment]
Hi @VargEM if you'd like to implement them I'd recommend taking a look at other distributions' implementations.  We're happy to help you write tests and documentation during the review process.

**Comment by VargEM** [comment]
And another question, how can I convert pytorch distribution to pyro distribution ??

**Comment by fehiepsi** [comment]
@VargEM I think you just need to make it a subclass of [TorchDistribution](https://github.com/pyro-ppl/pyro/blob/43c037443a05223b27e945fea26b2d4c2a0d79d8/pyro/distributions/polya_gamma.py#L13).

**Comment by VargEM** [comment]
I think this is the correct Johnson's SU distribution (correctness verified by comparing to scipy johnson's su dist):
I implemented it using 
https://reference.wolfram.com/language/ref/JohnsonDistribution.html
formulas

`from numbers import Number
import torch
from torch.distributions import constraints
from torch.distributions.distribution import Distribution
from torch.distributions.utils import broadcast_all
from torch.distributions import normal
import math
from pyro.distributions.torch_distribution import TorchDistributionMixin

class JohnsonSU(Distribution):
    r"""
    Creates a JohnsonSU distribution parameterized by :attr:`gamma` and :attr:`delta` and :attr:`mu`  and :attr:`sigma`.

    Example::

        >>> m = JohnsonSU(torch.tensor([2.0]), torch.tensor([8.0]), torch.tensor([4.0]), torch.tensor([1.0]), validate_args=True)
        >>> m.sample()
        tensor([3.8110])

    Args:
        gamma (float or Tensor): shape of the distribution
        delta (float or Tensor): shape of the distribution
        mu (float or Tensor): location of the distribution
        sigma (float or Tensor): scale of the distribution
    """

    arg_constraints = {'gamma': constraints.real, 'delta': constraints.positive,
                       'mu': constraints.real, 'sigma': constraints.positive}
    support = constraints.real

    @property
    def mean(self):
        rhs = torch.mul(torch.mul(torch.exp(torch.div(1, 2 * torch.pow(self.delta, 2))),
                                  self.sigma),
                        torch.sinh(torch.div(self.gamma, self.delta)))
        return torch.sub(self.mu, rhs)

    @property
    def variance(self):
        first_term = torch.div(torch.exp(torch.div(-2 * self.gamma, self.delta)), 4)

        second_term = torch.sub(torch.exp(torch.div(1, torch.pow(self.delta, 2))), 1)

        third_term = torch.add(torch.add(torch.exp(torch.div(1, torch.pow(self.delta, 2))),
                                         2 * torch.exp(torch.div(2 * self.gamma, self.delta))),
                               torch.exp(torch.div(1 + 4 * torch.mul(self.gamma, self.delta),
                                                   torch.pow(self.delta, 2))))
        return torch.mul(first_term,
                         torch.mul(second_term,
                                   torch.mul(third_term,
                                             torch.pow(self.sigma, 2))))

    @property
    def stddev(self):
        return torch.sqrt(self.variance)

    @property
    def median(self):
        return torch.add(self.mu, torch.mul(self.sigma, torch.sinh(torch.div(-self.gamma, self.delta))))

    def __init__(self, gamma, delta, mu, sigma, validate_args=None):
        self.gamma, self.delta, self.mu, self.sigma = broadcast_all(gamma, delta, mu, sigma)
        if isinstance(gamma, Number) and isinstance(delta, Number) and isinstance(mu, Number) and isinstance(sigma,
                                                                                                             Number):
            batch_shape = torch.Size()
        else:
            batch_shape = self.mu.size()
        super(JohnsonSU, self).__init__(batch_shape, validate_args=validate_args)

    def expand(self, batch_shape, _instance=None):
        new = self._get_checked_instance(JohnsonSU, _instance)
        batch_shape = torch.Size(batch_shape)
        new.gamma = self.gamma.expand(batch_shape)
        new.delta = self.delta.expand(batch_shape)
        new.mu = self.mu.expand(batch_shape)
        new.sigma = self.sigma.expand(batch_shape)
        super(JohnsonSU, new).__init__(batch_shape, validate_args=False)
        new._validate_args = self._validate_args
        return new

    def rsample(self, sample_shape=torch.Size()):
        shape = self._extended_shape(sample_shape)
        U = torch.rand(shape)
        return torch.add(torch.mul(self.sigma,
                                   torch.sinh(torch.div(torch.sub(normal.Normal(0, 1).icdf(U),
                                                                  self.gamma), self.delta))), self.mu)

    def log_prob(self, value):
        pi = torch.Tensor([math.pi])
        numerator = self.delta * torch.exp((-1 / 2) * (torch.pow(
            torch.add(self.gamma, torch.mul(self.delta, torch.arcsinh(torch.div(value - self.mu, self.sigma)))), 2)))
        denominator = torch.mul(torch.sqrt(2 * pi),
                                torch.sqrt(torch.pow(value - self.mu, 2) + torch.pow(self.sigma, 2)))
        return torch.log(numerator / denominator)

    def cdf(self, value):
        return (1 + torch.erf(
            torch.div(self.gamma + torch.mul(self.delta, torch.arcsinh(torch.div(value - self.mu, self.sigma))),
                      torch.sqrt(torch.tensor([2.0]))))) / 2
`



I tried to convert it to pyro dist using:
`class Johnson(JohnsonSU, TorchDistributionMixin):
    pass`

I want to use it in a simple MNIST VAE, and I changed the code structure to a 4 parameter distribution with loss=TraceGraph_ELBO(), but I always face these warnings:

/usr/local/lib/python3.7/dist-packages/pyro/poutine/trace_struct.py:227: UserWarning: Encountered NaN: log_prob_sum at site 'latent'
  warn_if_nan(site["log_prob_sum"], "log_prob_sum at site '{}'".format(name))
/usr/local/lib/python3.7/dist-packages/pyro/poutine/trace_struct.py:227: UserWarning: Encountered NaN: log_prob_sum at site 'obs'
  warn_if_nan(site["log_prob_sum"], "log_prob_sum at site '{}'".format(name))
/usr/local/lib/python3.7/dist-packages/pyro/poutine/trace_struct.py:255: UserWarning: Encountered NaN: log_prob_sum at site 'latent'
  warn_if_nan(site["log_prob_sum"], "log_prob_sum at site '{}'".format(name))
/usr/local/lib/python3.7/dist-packages/pyro/infer/tracegraph_elbo.py:266: UserWarning: Encountered NaN: loss
  warn_if_nan(loss, "loss")

What can I do??

**Comment by fritzo** [comment]
@VargEM I haven't looked into details, but in many distributions we've needed to clamp values to ensure numerical stability of both the forward and backward computations.  You'll probably need to play around, but you might start with something like
```py
    def __init__(self, gamma, delta, mu, sigma, validate_args=None):
        self.gamma, self.delta, self.mu, self.sigma = broadcast_all(gamma, delta, mu, sigma)
        finfo = torch.finfo(self.delta.dtype)
        self.delta = self.delta.clamp(min=finfo.eps)
        self.sigma = self.sigma.clamp(min=finfo.eps)
        batch_shape = self.mu.size()
        super(JohnsonSU, self).__init__(batch_shape, validate_args=validate_args)
```
